src/compression/b44/mod.rs

```diff
@@ -0,0 +1,924 @@
+                };
+
+                // for _ in 0..byte_count {
+                //     out.push(0);
+                // }
+
+                assert!(byte_count > 0);
+                assert!(byte_count <= 8);
+                assert_eq!(byte_count % 2, 0);
+                // println!("byte_count {}", byte_count);
+                // println!("row0 {}", row0);
+                // println!("out {}", out.len());
+                // println!("? {}", (row0 + byte_count) > out.len());
+                let sample_count = byte_count / 2;
+
+                if y + 3 < cd_ny {
+                    // cpy_u16_to_u8(&s[0..4], &mut out[row0..(row0 + byte_count)], byte_count);
+                    // cpy_u16_to_u8(&s[4..8], &mut out[row1..(row1 + byte_count)], byte_count);
+                    // cpy_u16_to_u8(&s[8..12], &mut out[row2..(row2 + byte_count)], byte_count);
+                    // cpy_u16_to_u8(&s[12..16], &mut out[row3..(row3 + byte_count)], byte_count);
+
+                    // println!("byte_count {}", byte_count);
+                    // println!("sample_count {}", sample_count);
+                    // println!("row0 {}", row0);
+                    cpy_u16(&s, 0, &mut tmp_u16_buffer, row0, sample_count);
+                    cpy_u16(&s, 4, &mut tmp_u16_buffer, row1, sample_count);
+                    cpy_u16(&s, 8, &mut tmp_u16_buffer, row2, sample_count);
+                    cpy_u16(&s, 12, &mut tmp_u16_buffer, row3, sample_count);
+
+                    // memcpy (row0, &s[ 0], n);
+                    // memcpy (row1, &s[ 4], n);
+                    // memcpy (row2, &s[ 8], n);
+                    // memcpy (row3, &s[12], n);
+                    // for i in 0..4 {
+                    //     u16::write(s[i + 0], &mut out[(row0 + i)..(row0 + i + 1)])
+                    //         .expect("write to in-memory failed");
+                    //     u16::write(s[i + 4], &mut out[(row1 + i)..(row1 + i + 1)])
+                    //         .expect("write to in-memory failed");
+                    //     u16::write(s[i + 8], &mut out[(row2 + i)..(row2 + i + 1)])
+                    //         .expect("write to in-memory failed");
+                    //     u16::write(s[i + 12], &mut out[(row3 + i)..(row3 + i + 1)])
+                    //         .expect("write to in-memory failed");
+                    //     // out[(row0 + i)..(row0 + i + 1)] = s[i + 0].to_be_bytes();
+                    //     // out[(row1 + i)..(row1 + i + 1)] = s[i + 4].to_be_bytes();
+                    //     // out[(row2 + i)..(row2 + i + 1)] = s[i + 8].to_be_bytes();
+                    //     // out[(row3 + i)..(row3 + i + 1)] = s[i + 12].to_be_bytes();
+                    //     // out[row0 + i] = s[i + 0];
+                    //     // out[row1 + i] = s[i + 4];
+                    //     // out[row2 + i] = s[i + 8];
+                    //     // out[row3 + i] = s[i + 12];
+                    // }
```

Comment:
I'm sure there is a better way to do this. Any advice is welcome. What could be the proper way to do `memcpy (row0, &s[ 0], n);` ?

Reply:
I think I would replace the indexing with sub slices. By the way, I think you might have mixed up the source and destination?

For example:

```rust
// cpy_u16(&s, 0, &mut tmp_u16_buffer, row0, sample_count);
// instead, I might do:
row0: &mut [u16] = ...;
row1: &mut [u16] = ...;

row0.copy_from_slice(&s[0 .. 4]);
row1.copy_from_slice(&s[4 .. 8]); // or [ 4 .. 4 + sample_count ] ?
row2.copy_from_slice(&s[8 .. 12]);
row3.copy_from_slice(&s[12 .. 16]);
```

However, this would create a borrowing issue, as the different rows probably all come from one array, and you can't simply borrow multiple mutable slices from the same array. There are two possibilities:

A: Borrow the individual rows on the fly every time you want to do anything:
```rust
all_rows[row1_start .. row1_end].copy_from_slice(&s[0 .. 4]);
all_rows[row2_start .. row2_end].copy_from_slice(&s[4 .. 8]); // or [ 4 .. 4 + sample_count ] ?
all_rows[row3_start .. row3_end].copy_from_slice(&s[8 .. 12]);
all_rows[row4_start .. row4_end].copy_from_slice(&s[12 .. 16]);
```

B: My preferred solution: Use slices all the way, avoiding indices wherever possible.
```rust
// two borrow multiple slices out of the same buffer, use one of these approaches:
let (row0, buffer_rest) = buffer.split_mut(row0_end); // assuming all rows immediately follow each other in the buffer
let (row1, buffer_rest) = buffer_rest.split_mut(row1_end); // you can also call this on a subslice of the buffer_rest
let (row2, buffer_rest) = buffer_rest.split_mut(row2_end);
...
```

```rust
// split buffer into multiple slices:
let mut remaining_rows_iter = buffer.chunks_exact_mut(row_size); // split into parts, assuming all rows have the same size 
for (row, index) in remaining_rows.enumerate() { row.copy_from_slice(&s[index * 4 .. index * 4 + count]); }

// you can also extract the iter rows manually without the for loop
let row0 = remaining_rows_iter.next().unwrap();
let row1 = remaining_rows_iter.next().unwrap();
let row2 = remaining_rows_iter.next().unwrap();
```

Reply:
I have multiple troubles with this one.

As B44 is decompressing 4x4 blocks, it have to be initialized once so you can write 4x4 pixels in the buffer. This mean you can't `push()` on it easily. You have to initialize it and write 4x4 blocks on it.

I'm digging with `copy_from_slice()` because I want `memcpy`. I realize there are no proper way to convert a `u16` slice to a `u8` one.

```rust
    let s_slice = &s[s_i..(s_i+n)];  // &[u16]
    let out_slice = &mut out[(row*2)..((row*2)+n*2)];  // &[u8]

    out_slice.copy_from_slice(s_slice); // fail to compile.
```

I suspect the only way to do a real `memcpy` like [the reference implementation](https://github.com/AcademySoftwareFoundation/openexr/blob/master/src/lib/OpenEXR/ImfB44Compressor.cpp#L950) is to have a `unsafe` dedicated to 4 u16 copy to u8 slice. All others will imply a unrolled loop. I can do a loop, we simply miss the `memcpy`. You are the maintainer, you choose. :)

```rust
    let mut out_i = row*2;

    for v in &s[s_i..(s_i+n)] {
        let v_array = v.to_be_bytes();
        out[out_i] = v_array[0];
        out[out_i+1] = v_array[1];
        out_i +=2;
    }
```

Reply:
This is exactly the purpose of the `lebe` library. You can not only use it to convert endianness, but also convert between bytes and u16 slices without changing endianness. 

See here: https://docs.rs/lebe/0.5.1/lebe/io/trait.WriteEndian.html#method.write_as_native_endian

To use this writer method, you can call `std::io::Cursor::new(&mut slice)` to create a writer, which can then overwrite the contents in the `slice` using any IO operation. One kind of IO operation would be to convert endianness, or another would be to simply copy the contents of another slice.
 
Full (untested) example:

```rust
use lebe::prelude::*;
use std::io::Cursor;

fn convert_u16_to_u8(){
    let mut bytes = vec![0_u8; 1000]; // allocate 1000 bytes output
    let u16_values = vec![ 2343; 32 ]; // our 32 numbers 
    
    let mut bytes_writer = Cursor::new(&mut bytes[2 .. ]); // overwrite the contents of the byte vector, starting at the third byte
    bytes_writer.write_as_native_endian(u16_values).unwrap(); // overwrite 32*sizeof(u16) bytes, internally using a memcpy, without converting endianness, the first three bytes are not touched
}

```

Reply:
Thanks, I looked a little into the `lebe` libs, I missed the `Cursor`, but it still worry me. Does not `Cursor::new()` allocate on the heap?

I cant test on playground rust, but I suspect this `convert_u16_to_u8()` will not be compiled to `memcpy` instruction only because of the heap allocation. Allocating on heap kinda lose the benefit of `memcpy` IMHO.

I also suspect `lebe` to focus on write/read on/from IO stream of byte. Here we are manipulating in-memory arrays/slice.

Reply:
> Does not Cursor::new() allocate on the heap?

No, I don't think so. Actually, the slice on its own already implements `Read`. Technically we don't need the cursor to use a vector like a file, but it delivers intent. It does not allocate if you pass it a slice. It only allocates if you pass it a vec and the vec needs to allocate.

> I also suspect lebe to focus on write/read on/from IO stream of byte. Here we are manipulating in-memory arrays/slice.

The purpose of lebe is twofold: A) to make converting between bytes and slices possible without the use of unsafe, B) be endian-aware while doing that.

The Rust `Read` and `Write` trait are not restricted to files. They are often used for in-memory manipulation by design. They will always compile to a memcpy in that situation. 

Doing conversion from u32/f32/f16 to bytes is the main goal of lebe, and it does not matter whether your data is in memory or being written to a physical file. :)

Reply:
After more investigation, I don't think I can skip the [use of a temp buffer](https://github.com/AcademySoftwareFoundation/openexr/blob/master/src/lib/OpenEXR/ImfB44Compressor.cpp#L866). It seems the b44 uncompress implementation first uncompress block in the order they are compressed. My efforts to remove the use of a temp buffer to directly write inside the output buffer would means a big move away from the reference implementation. I think it's too soon.

I suggest to keep the current implementation (use a u16 buffer, then convert to scanline u8 buffer), and continue to clean the code and don't touch to the implementation.

Reply:
Sounds good. Optimization should be the last step, if there is no simple solution. :)